### PR TITLE
udev: revert bind/unbind patch and one more

### LINKED ISSUE #11277: network interface not renamed under gentoo with openrc and netifrc [systemd 240]
**systemd version the issue has been seen with**
239, 240

**Used distribution**
Gentoo

**Expected behaviour you didn't see**
stable network interfaces: enp0s3 

**Unexpected behaviour you saw**
changing network interfaces: ens3 with udev-239, eth0 with udev-240

**Steps to reproduce the problem**
update udev from 238 to 239 and above

see also: https://bugs.gentoo.org/673360

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #11277, #11299, #11314, and #11319.

### CODE REVIEW

File: src/udev/udevd.c [hunk 1]
```diff
@@ -534,7 +534,7 @@ static int worker_spawn(Manager *manager, struct event *event) {
         if (r < 0)
                 return log_error_errno(r, "Worker: Failed to enable receiving of device: %m");
 
-        r = safe_fork("(worker)", FORK_DEATHSIG, &pid);
+        r = safe_fork(NULL, FORK_DEATHSIG, &pid);
```
Comment (yuwata) [inline src/udev/udevd.c:537]: BTW, I've tried to set the worker's process name, but it does not work. At least `journald` does not show process name correctly. `udevd` does not have enough privilege to change process name?

**Comment by keszybz** [comment]
So, without this patch, and with SYSTEMD_LOG_LEVEL=debug enabled, boot fails. My understanding is that this is a race, and with udevd slowed down a bit by the logging, we fairly reliably hit it. So the first revert makes sense.

But why the second revert?

**Comment by yuwata** [comment]
> the second revert

Because `journalctl` shows something strange...
I cannot start X now, so it is hard to paste relevant logs. Sorry.

**Comment by yuwata** [comment]
> My understanding is that this is a race, and with udevd slowed down a bit by the logging, we fairly reliably hit it.

With debug logging, udevd crashes so frequently. I guess that udevd does not handle error by sd-device correctly. I will take a look in detail tomorrow.

**Comment by yuwata** [comment]
@keszybz I've added several commits:
- re-implement ignoring bind/unbind events on udevd.
- fixes the origin of the segfault.
- add a test case, but it does not reproduce the segfault... But, anyway, adding more tests is not bad.

PTAL.

**Comment by yuwata** [comment]
Aha, finally I got X! Below is the original segfault when debug logging is enabled:
```
Dec 22 19:34:18 systemd-coredump[584]: Process 573 (systemd-udevd) of user 0 dumped core.
                                       
                                       Stack trace of thread 573:
                                       #0  0x00007f5fffc1186b __strrchr_avx2 (libc.so.6)
                                       #1  0x00007f5fff89dfa1 n/a (libsystemd-shared-240.so)
                                       #2  0x00007f5fff89e1a5 sd_device_get_sysname (libsystemd-shared-240.so)
                                       #3  0x00007f5fff8a7ea6 n/a (libsystemd-shared-240.so)
                                       #4  0x00007f5fff8a8038 device_new_from_nulstr (libsystemd-shared-240.so)
                                       #5  0x00007f5fff8a87d1 device_monitor_receive_device (libsystemd-shared-240.so)
                                       #6  0x00007f5fff8a8bf4 n/a (libsystemd-shared-240.so)
                                       #7  0x00007f5fff885540 n/a (libsystemd-shared-240.so)
                                       #8  0x00007f5fff88586d sd_event_dispatch (libsystemd-shared-240.so)
                                       #9  0x00007f5fff885a08 sd_event_run (libsystemd-shared-240.so)
                                       #10 0x00007f5fff885c3b sd_event_loop (libsystemd-shared-240.so)
                                       #11 0x000055baec0f0399 n/a (systemd-udevd)
                                       #12 0x00007f5fffad4413 __libc_start_main (libc.so.6)
                                       #13 0x000055baec0f121e n/a (systemd-udevd)
```
So, the story is,
1. device monitor receives a device with 'bind' or 'unbind' action.
2. `device_append()` ignores the action and try to logs about that.
3. However, `devpath` may not be set yet for the device.
4. So, `log_device_*()` causes segfault as it calls `sd_device_get_sysname()` when debug logging is enabled.

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging 2b10b668de4c44808a95354cc29f3d990f87b920 into 8eab766804ef4fa21d26c00fd0baab3f1a47bb5c - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-fd7a0b53b67138ec62e11c41d9e7c36af16dcf26)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by evverx** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 8d83a62dba32be16736eb378289a8f8b48691a97 into 04ca4d191b13e79e5701ed22dc972f08628d7bcc - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-ce3c27527a8f634da5d4b8573d6f442673a9abef)

**new alerts:**

* 1 for FIXME comment

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging 4fe0caadc85431118f2d8aea7570307cfc2aed27 into 04ca4d191b13e79e5701ed22dc972f08628d7bcc - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-8abe389f1600b803a2a6a09c8174788fc8e3cfcd)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by iv-m** [comment]
I guess  56c886d should not be reverted, since the crash is fixed by 18fee12.

**Comment by yuwata** [comment]
It seems dropping bind/unbind actions causes may issues, #11277, #11299, #11314.

**Comment by yuwata** [comment]
@keszybz and @poettering 56c886dc7ed5b2bb0882ba85136f4070545bfc1b (#11101) may fixes something (#7587? I am not sure) causes so many new issues... So, I would like to once revert the commit, and implement in a better way later. What do you think?

At least, such the logic should not be implemented in sd-device, but in udevd.

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging 4fe0caadc85431118f2d8aea7570307cfc2aed27 into aa70783f55b369521b94e0985e84bbdaae16b174 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-6fcaeeebefa6ff7106ad4aac384019fee2b1dd8f)

**fixed alerts:**

* 1 for FIXME comment

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by keszybz** [comment]
Yep, I think we should merge this. #7587 is reintroduced, but the fix in #11101 was causing too many issues.

**Pull request merged**